aspnetcore/blazor/Security-guidance.md

```diff
@@ -0,0 +1,254 @@
+* Don't trust data on event handlers before validating it first.
+* Take appropriate action upon receiving what your application considers to be invalid data.
+  * Ignore the data and return, if you can't be sure the input was illegitimate. This will allow the app to continue working.
+  * Throw an exception, if you determine that the input is illegitimate and couldn't have been produced by legitimate clients. This will tear down the circuit and end the session.
+* Don't trust the error message comming from render batch completions included in the logs. This error came from the client and can't be generally trusted as the client might have been compromised.
+* Don't trust the input from interop calls. This input comes in two flavors.
+  * JS to .NET interop invocations.
+  * .NET to JS interop results.
+* The application is responsible for validating that the contents of the arguments/result are valid even if they were correctly deserialized.
+
+Blazor server-side takes specific steps to mitigate some of these threats.
+* It will stop producing new UI updates if the client is not acknowledging render batches. (This can be configured in `CircuitOptions.MaxBufferedUnacknowledgedRenderBatches`)
+* It will timeout any .NET to JS call after 1 minute without receiving a response from the client.
+* It performs some basic validation in all input comming from the browser from JS interop.
+  * Ensure that the dotnet references are valid and of the type expected by the .NET method.
+  * Ensure that the data is not malformed.
+  * Ensure the right number of arguments for the method is present on the payload.
+  * Ensure that the arguments/result can be deserialized correctly before invoking the method.
+* It performs some basic validation in all input comming from the browser from dispatched events.
+  * It validates that the event has a valid type.
+  * It validates that the data for the event can be deserialized.
+  * It validates that there is an event handler associated with that event.
+
+For a XSS vulnerability to be possible, the application must allow some user input to be included as part of the rendered page. Blazor server-side components go through a compile-time step where the contents of the page are transformed into what we call a RenderTree (a virtual DOM like structure) and the diffs between previous and current renders are sent to the client for rendering, or in the case of prerendering serialized into HTML.
+* At this step blazor server-side takes care of HTML encoding all content that wasn't defined as static markup at compile-time.
+* Script tags aren't allowed to be included as components and a compile time error is generated when one such tag is included in a razor page.
+  * Script tags will not behave the way one would expect on the page. Once added they execute the contents of the script and those side-effects don't disappear even if you remove the script tag.
```

Comment:
```suggestion
  * The reason that script tags are disallowed is because they would not behave the way one would expect on the page. Once added they execute the contents of the script and those side-effects don't disappear even if you remove the script tag.
```